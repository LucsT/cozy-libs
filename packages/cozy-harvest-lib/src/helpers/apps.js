const STORE_SLUG = 'store'

/**
 * Return Store URL where an app/konnector can be installed / updated
 * @param  {Array}  [appData=[]]   Apps data, as returned by endpoint /apps/ or
 * /konnectors/
 * @param  {Object} [app={}] AppObject
 * @return {String}                URL as string
 */
export const getStoreUrltoInstallAnApp = (appData = [], app = {}) => {
  if (!app.slug) {
    throw new Error('Expected app / konnector with defined slug')
  }

  const storeApp = isAppInstalled(appData, { slug: STORE_SLUG })
  if (!storeApp) return null

  const storeUrl = storeApp.links && storeApp.links.related

  if (!storeUrl) return null

  return `${storeUrl}#/discover/${app.slug}/install`
}

/**
 *
 * @param {Array} apps Array of apps returned by /apps /konnectors
 * @param {Object} wantedApp io.cozy.app with at least a slug
 * @return {Object} The io.cozy.app is installed or undefined if not
 */
export const isAppInstalled = (apps = [], wantedApp = {}) => {
  return apps.find(
    app => app.attributes && app.attributes.slug === wantedApp.slug
  )
}
/**
 *
 * @param {Object} app io.cozy.app object
 * @return {String} url to the app
 */
export const getUrlForApp = app => {
  return app.links && app.links.related
}
export default {
  getStoreUrltoInstallAnApp
}
