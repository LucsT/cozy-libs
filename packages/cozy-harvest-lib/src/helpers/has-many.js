import get from 'lodash/get'
import merge from 'lodash/merge'

const updateArray = (array, index, el) => {
  return [...array.slice(0, index), el, ...array.slice(index + 1)]
}

export const getHasManyItem = (doc, relationshipName, relationshipItemId) => {
  const relationships = get(doc, `relationships.${relationshipName}.data`, [])
  return relationships.find(rel => rel._id == relationshipItemId)
}

export const setHasManyItem = (
  document,
  relationshipName,
  relId,
  relAttributes
) => {
  const relationship = get(
    document,
    `relationships.${relationshipName}.data`,
    []
  )
  const relIndex = relationship.findIndex(rel => rel._id === relId)
  const updatedRel = merge({}, relationship[relIndex], relAttributes)
  const updatedRelationship = updateArray(relationship, relIndex, updatedRel)
  const updatedDocument = {
    ...document,
    relationships: {
      ...(document.relationships || {}),
      [relationshipName]: {
        data: updatedRelationship
      }
    }
  }
  return updatedDocument
}

export const updateHasManyItem = (
  document,
  relationshipName,
  relId,
  updater
) => {
  const relItem = getHasManyItem(document, relationshipName, relId)
  const updatedRelItem = updater(relItem)
  return setHasManyItem(document, relationshipName, relId, updatedRelItem)
}
