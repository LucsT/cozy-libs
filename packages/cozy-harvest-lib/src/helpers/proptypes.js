import PropTypes from 'prop-types'

function getMutationsProptypes(mutations) {
  return {
    ...Object.keys(mutations()).reduce((propTypes, mutationName) => {
      propTypes[mutationName] = PropTypes.func
      return propTypes
    }, {}),
    // FIXME use directly proptypes from cozy-client when available
    createDocument: PropTypes.func,
    saveDocument: PropTypes.func,
    deleteDocument: PropTypes.func
  }
}

export { getMutationsProptypes }
