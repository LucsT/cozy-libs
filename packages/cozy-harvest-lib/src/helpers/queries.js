import CozyClient, { Q } from 'cozy-client'

const DEFAULT_CACHE_TIMEOUT_QUERIES = 1 * 60 * 1000
const defaultFetchPolicy = CozyClient.fetchPolicies.olderThan(
  DEFAULT_CACHE_TIMEOUT_QUERIES
)

export const buildKonnectorQuery = (slug, source) => {
  const query = {
    definition: () =>
      Q('io.cozy.konnectors').getById(`io.cozy.konnectors/${slug}`),
    options: {
      as: `konnector-${slug}-${source}`,
      fetchPolicy: defaultFetchPolicy
    }
  }
  const queryDef = query.definition()
  queryDef.sources = [source]

  return { definition: queryDef, options: query.options }
}
