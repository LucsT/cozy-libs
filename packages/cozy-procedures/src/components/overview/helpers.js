import formatDate from 'date-fns/format'
import subMonths from 'date-fns/sub_months'
import startOfMonth from 'date-fns/start_of_month'
import pick from 'lodash/pick'
import groupBy from 'lodash/groupBy'

/**
 * createFileWithContent - upload a json file containing the given data
 *
 * @param {string} filename - The file name
 * @param {string} destinationId - The id of the destination directory
 * @param {string} content - The file content
 * @return {object} Information about the file (id and name)
 */
export const createFileWithContent = async (
  client,
  filename,
  destinationId,
  content
) => {
  const file = new File([content], filename)
  const resp = await client
    .collection('io.cozy.files')
    .createFile(file, { dirId: destinationId })

  return resp.data
}

export const fetchCheckingsLikeAccounts = client => {
  const query = client.find('io.cozy.bank.accounts').where({
    type: {
      $in: ['Checkings', 'Bank', 'Cash', 'Deposit']
    }
  })

  return client.queryAll(query)
}

export const fetchTransactionsHistory = async (client, accountsIds) => {
  const today = startOfMonth(new Date())
  const threeMonthsBefore = startOfMonth(subMonths(today, 3))

  const query = client.find('io.cozy.bank.operations').where({
    account: { $in: accountsIds },
    date: {
      $lt: formatDate(today, 'YYYY-MM-DD'),
      $gt: formatDate(threeMonthsBefore, 'YYYY-MM-DD')
    }
  })

  return client.queryAll(query)
}

export const fetchTransactionsHistoryContent = async client => {
  const accounts = await fetchCheckingsLikeAccounts(client)
  const accountsIds = accounts.map(account => account._id)
  const transactions = await fetchTransactionsHistory(client, accountsIds)

  const transactionsByAccount = groupBy(
    transactions,
    transaction => transaction.account
  )

  const rawContent = accounts.map(account => {
    return {
      ...pick(account, ['balance', 'iban']),
      transactions: transactionsByAccount[account._id].map(transaction =>
        pick(transaction, ['amount', 'label', 'date'])
      )
    }
  })

  return JSON.stringify(rawContent)
}
