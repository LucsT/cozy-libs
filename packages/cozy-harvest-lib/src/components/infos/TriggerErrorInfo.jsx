import React, { PureComponent } from 'react'
import PropTypes from 'prop-types'

import { translate } from 'cozy-ui/transpiled/react/I18n'
import Infos from 'cozy-ui/transpiled/react/Infos'
import Typography from 'cozy-ui/transpiled/react/Typography'

import { getErrorLocale } from '../../helpers/konnectors'
import withKonnectorLocales from '../hoc/withKonnectorLocales'
import Markdown from '../Markdown'

/**
 * Component used to show an error related to an account or to a konnector job
 * Regular error are treated in a common way, but for KonnectorJobErrors
 * we can add a little bit of logic.
 *
 * This component is strongly related with the content of locales files and
 * deals mainly with translation concerns.
 */
export class TriggerErrorInfo extends PureComponent {
  render() {
    const { className, error, konnector, t } = this.props
    return (
      <Infos
        className={className}
        theme="danger"
        description={
          <>
            <Typography className="u-error" variant="h6" gutterBottom>
              {getErrorLocale(error, konnector, t, 'title')}
            </Typography>
            <Typography variant="body1" component="div">
              <Markdown
                source={getErrorLocale(error, konnector, t, 'description')}
              />
            </Typography>
          </>
        }
      />
    )
  }
}

TriggerErrorInfo.proptTypes = {
  error: PropTypes.object.isRequired,
  konnector: PropTypes.object.isRequired,
  t: PropTypes.func.isRequired
}

export default translate()(withKonnectorLocales(TriggerErrorInfo))
