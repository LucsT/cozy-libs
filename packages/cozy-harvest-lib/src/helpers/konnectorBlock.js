import { generateUniversalLink } from 'cozy-ui/transpiled/react/AppLinker'

import { buildKonnectorQuery } from './queries'

const fetchAndFormatKonnector = async ({
  client,
  query,
  universalLink,
  iconStatus
}) => {
  try {
    await client.query(query.definition, query.options)
    const res = client.getQueryFromState(query.options.as)
    const { data } = res
    const konnector = data[0].attributes

    const link = generateUniversalLink({
      cozyUrl: client.getStackClient().uri,
      slug: universalLink.slug,
      subDomainType: client.getInstanceOptions().cozySubdomainType,
      nativePath: universalLink.nativePath
    })

    return {
      name: konnector.name,
      link,
      vendorLink: konnector.vendor_link && {
        component: 'a',
        href: konnector.vendor_link,
        target: '_blank'
      },
      iconStatus
    }
  } catch (error) {
    throw error
  }
}

export const fetchKonnectorFromStackOrRegistry = async ({
  client,
  slug,
  sourceAccount
}) => {
  try {
    return await fetchAndFormatKonnector({
      client,
      query: buildKonnectorQuery(slug, 'stack'),
      universalLink: {
        slug: 'home',
        nativePath: `connected/${slug}/accounts/${sourceAccount}`
      }
    })
  } catch (error) {
    // The konnector could be uninstalled
    if (error.status === 404) {
      try {
        return await fetchAndFormatKonnector({
          client,
          query: buildKonnectorQuery(slug, 'registry'),
          universalLink: {
            slug: 'store',
            nativePath: `discover/${slug}`
          },
          iconStatus: 'disabled'
        })
      } catch (error) {
        throw error
      }
    }
  }
}
