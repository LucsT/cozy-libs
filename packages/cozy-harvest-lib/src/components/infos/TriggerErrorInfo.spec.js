import React from 'react'
import { shallow } from 'enzyme'

import { TriggerErrorInfo } from 'components/infos/TriggerErrorInfo'
import { KonnectorJobError } from 'helpers/konnectors'

const fixtures = {
  konnector: {
    name: 'Konnectest',
    vendor_link: 'https://cozy.io'
  },
  error: new KonnectorJobError('LOGIN_FAILED')
}

const tMock = jest.fn().mockImplementation(key => key)

describe('TriggerErrorInfo', () => {
  beforeAll(() => {
    jest.spyOn(console, 'error').mockImplementation(() => {})
  })

  afterAll(() => {
    jest.restoreAllMocks()
  })

  const props = {
    error: fixtures.error,
    konnector: fixtures.konnector,
    t: tMock
  }

  it('should render', () => {
    const component = shallow(<TriggerErrorInfo {...props} />).getElement()
    expect(component).toMatchSnapshot()
  })

  it('should render regular Error', () => {
    const component = shallow(
      <TriggerErrorInfo
        {...props}
        error={new Error('Something is undefined')}
      />
    ).getElement()
    expect(component).toMatchSnapshot()
  })
})
