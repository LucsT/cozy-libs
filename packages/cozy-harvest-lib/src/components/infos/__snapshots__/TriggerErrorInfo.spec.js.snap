// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TriggerErrorInfo should render 1`] = `
<InfosMigration
  description={
    <React.Fragment>
      <WithStyles(Typography)
        className="u-error"
        gutterBottom={true}
        variant="h6"
      >
        error.job.LOGIN_FAILED.title
      </WithStyles(Typography)>
      <WithStyles(Typography)
        variant="body1"
      >
        <Markdown
          source="error.job.LOGIN_FAILED.description"
        />
      </WithStyles(Typography)>
    </React.Fragment>
  }
  theme="danger"
/>
`;

exports[`TriggerErrorInfo should render regular Error 1`] = `
<InfosMigration
  description={
    <React.Fragment>
      <WithStyles(Typography)
        className="u-error"
        gutterBottom={true}
        variant="h6"
      >
        error.job.UNKNOWN_ERROR.title
      </WithStyles(Typography)>
      <WithStyles(Typography)
        variant="body1"
      >
        <Markdown
          source="error.job.UNKNOWN_ERROR.description (Something is undefined)"
        />
      </WithStyles(Typography)>
    </React.Fragment>
  }
  theme="danger"
/>
`;
